"This store is closed" after two-phase commit

Please be advised that there is some issue wrt. two-phase commit in the current master.
The issue manifests itself in the following way: after a 2pc operation the DB seems to be working normally: read, insert ops are fine. But when I call 'dbConnection.close' the following entries appear in the tracefile:

	...
	Caused by: org.h2.jdbc.JdbcSQLNonTransientException: IO Exception: "Closing" [90028-219]
		at org.h2.message.DbException.getJdbcSQLException(DbException.java:554)
		at org.h2.message.DbException.getJdbcSQLException(DbException.java:477)
		... 34 more
	Caused by: org.h2.mvstore.MVStoreException: This store is closed [2.2.219/4]
		at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996)
		at org.h2.mvstore.MVStore.checkOpen(MVStore.java:1426)
		at org.h2.mvstore.MVStore.getMetaMap(MVStore.java:561)
		at org.h2.mvstore.FileStore.rewriteChunks(FileStore.java:1902)
		at org.h2.mvstore.FileStore.compactRewrite(FileStore.java:1887)
		at org.h2.mvstore.FileStore.rewriteChunks(FileStore.java:1848)
		at org.h2.mvstore.FileStore.lambda$compact$0(FileStore.java:876)
		at org.h2.mvstore.MVStore.tryExecuteUnderStoreLock(MVStore.java:947)
		at org.h2.mvstore.FileStore.compact(FileStore.java:876)
		at org.h2.mvstore.RandomAccessStore.compactStore(RandomAccessStore.java:425)
		at org.h2.mvstore.FileStore.compactStore(FileStore.java:887)
		at org.h2.mvstore.FileStore.stop(FileStore.java:275)
		at org.h2.mvstore.MVStore.closeStore(MVStore.java:692)
		at org.h2.mvstore.MVStore.close(MVStore.java:643)
		at org.h2.mvstore.db.Store.close(Store.java:364)
		... 32 more


This issue is present with both BLOB and VARBINARY types. There was no such problem on H2-2.1.214 and earlier versions.

--- a/h2/src/main/org/h2/mvstore/Chunk.java
+++ b/h2/src/main/org/h2/mvstore/Chunk.java
@@ -36,7 +36,7 @@ public abstract class Chunk<C extends Chunk<C>> {
      * root:ffffffffffffffff,time:ffffffffffffffff,toc:ffffffff,version:ffffffffffffffff,
      * next:ffffffffffffffff
      */
-    static final int MAX_HEADER_LENGTH = 178;
+    static final int MAX_HEADER_LENGTH = 1024;
 
     /**
      * The length of the chunk footer. The longest footer is:

--- a/h2/src/main/org/h2/mvstore/FileStore.java
+++ b/h2/src/main/org/h2/mvstore/FileStore.java
@@ -1437,7 +1437,7 @@ public abstract class FileStore<C extends Chunk<C>>
     private void serializeToBuffer(WriteBuffer buff, ArrayList<Page<?, ?>> changed, C c, C previousChunk) {
         // need to patch the header later
         c.writeChunkHeader(buff, 0);
-        int headerLength = buff.position() + 66; // len:0[fffffff]map:0[fffffff],toc:0[fffffffffffffff],root:0[fffffffffffffff,next:ffffffffffffffff]
+        int headerLength = buff.position() + 58; // len:0[fffffff]map:0[fffffff],toc:0[ffffffff],root:0[fffffffffffffff,next:ffffffffffffffff]
         buff.position(headerLength);
         c.next = headerLength;
 

--- a/h2/src/main/org/h2/mvstore/MVStore.java
+++ b/h2/src/main/org/h2/mvstore/MVStore.java
@@ -494,7 +494,7 @@ public class MVStore implements AutoCloseable {
      * @return Map
      */
     public <K, V> MVMap<K,V> getMap(int id) {
-        checkOpen();
+        checkNotClosed();
         @SuppressWarnings("unchecked")
         MVMap<K, V> map = (MVMap<K, V>) maps.get(id);
         return map;
@@ -507,7 +507,7 @@ public class MVStore implements AutoCloseable {
      */
     public Set<String> getMapNames() {
         HashSet<String> set = new HashSet<>();
-        checkOpen();
+        checkNotClosed();
         for (Iterator<String> it = meta.keyIterator(DataUtils.META_NAME); it.hasNext();) {
             String x = it.next();
             if (!x.startsWith(DataUtils.META_NAME)) {
@@ -558,7 +558,7 @@ public class MVStore implements AutoCloseable {
      * @return the metadata map
      */
     public MVMap<String, String> getMetaMap() {
-        checkOpen();
+        checkNotClosed();
         return meta;
     }
 
@@ -928,7 +928,7 @@ public class MVStore implements AutoCloseable {
     public void executeFilestoreOperation(Runnable operation) {
         storeLock.lock();
         try {
-            checkOpen();
+            checkNotClosed();
             fileStore.executeFilestoreOperation(operation);
         } catch (MVStoreException e) {
             panic(e);

